data/transactions/logic/mocktracer/tracer.go

```diff
@@ -0,0 +1,147 @@
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package mocktracer
+
+import (
+	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/data/transactions/logic"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+// EventType represents a type of logic.EvalTracer event
+type EventType string
+
+const (
+	// BeforeTxnGroupEvent represents the logic.EvalTracer.BeforeTxnGroup event
+	BeforeTxnGroupEvent EventType = "BeforeTxnGroup"
+	// AfterTxnGroupEvent represents the logic.EvalTracer.AfterTxnGroup event
+	AfterTxnGroupEvent EventType = "AfterTxnGroup"
+	// BeforeTxnEvent represents the logic.EvalTracer.BeforeTxn event
+	BeforeTxnEvent EventType = "BeforeTxn"
+	// AfterTxnEvent represents the logic.EvalTracer.AfterTxn event
+	AfterTxnEvent EventType = "AfterTxn"
+	// BeforeProgramEvent represents the logic.EvalTracer.BeforeProgram event
+	BeforeProgramEvent EventType = "BeforeProgram"
+	// AfterProgramEvent represents the logic.EvalTracer.AfterProgram event
+	AfterProgramEvent EventType = "AfterProgram"
+	// BeforeOpcodeEvent represents the logic.EvalTracer.BeforeOpcode event
+	BeforeOpcodeEvent EventType = "BeforeOpcode"
+	// AfterOpcodeEvent represents the logic.EvalTracer.AfterOpcode event
+	AfterOpcodeEvent EventType = "AfterOpcode"
+)
+
+// Event represents a logic.EvalTracer event
+type Event struct {
+	Type EventType
+
+	// only for BeforeProgram and AfterProgram
+	LogicEvalMode logic.RunMode
+
+	// only for BeforeTxn and AfterTxn
+	TxnType protocol.TxType
+
+	// only for AfterTxn
+	TxnApplyData transactions.ApplyData
+
+	// only for BeforeTxnGroup and AfterTxnGroup
+	GroupSize int
+}
+
+// BeforeTxnGroup creates a new Event with the type BeforeTxnGroupEvent
+func BeforeTxnGroup(groupSize int) Event {
+	return Event{Type: BeforeTxnGroupEvent, GroupSize: groupSize}
+}
+
+// AfterTxnGroup creates a new Event with the type AfterTxnGroupEvent
+func AfterTxnGroup(groupSize int) Event {
+	return Event{Type: AfterTxnGroupEvent, GroupSize: groupSize}
+}
+
+// BeforeProgram creates a new Event with the type BeforeProgramEvent
+func BeforeProgram(mode logic.RunMode) Event {
+	return Event{Type: BeforeProgramEvent, LogicEvalMode: mode}
+}
+
+// BeforeTxn creates a new Event with the type BeforeTxnEvent
+func BeforeTxn(txnType protocol.TxType) Event {
+	return Event{Type: BeforeTxnEvent, TxnType: txnType}
+}
+
+// AfterTxn creates a new Event with the type AfterTxnEvent
+func AfterTxn(txnType protocol.TxType, ad transactions.ApplyData) Event {
+	return Event{Type: AfterTxnEvent, TxnType: txnType, TxnApplyData: ad}
+}
```

Comment:
Won't these hooks want to inspect the txns?

I see AfterTxn gets the AD, but the other hook functions don't get to see the object in question, whether it's the group or individual transaction. Why is that?

Reply:
They do have access to the transactions. BeforeTxnGroup, AfterTxnGroup, BeforeTxn, and AfterTxn all receive a `logic.EvalParams` as an argument, from which they can access the `TxnGroup` field.

What you're looking at here are functions used to implement the mock tracer, so they don't actually look at the transactions, but any interesting tracer likely would.

Reply:
ugh, sorry. I thought I was looking at an implementation of the interface here, but of course not, these don't even have receivers.